### PR TITLE
Auto-balancing test runners

### PR DESCRIPTION
**Pull request by mpalmi**
Integrate the `test-go` workflow with `gotestsum tool ci-matrix`. The tool uses the output of `go list ./...` along with timing files emitted by `gotestsum` to generate a test matrix of 16 runners with evenly distributed runtimes.

We intentionally ignore binary, docker-based test files for the initial matrix creation and then inject a 17th runner, dedicated to building Vault and running the entire binary test suite together. This avoids duplication of build overhead when binary tests are rebalanced across multiple runners in the generated matrix.

In order to maintain test results from previous runs, we cache the test results after every run of `gotestsum`. Each cache entry occupies ~36MB after compression on enterprise, at the time of this commit.

We'll have to keep an eye on this to make sure timing data is not evicted from the cache, but in theory it should be toward the top of the LRU entries.

Resolves VAULT-17584

### CODE REVIEW

**Comment by mpalmi** [comment]
We should get a cache hit on the second run of CI (assuming I didn't mess something up). I'll rerun all the workflows once the tests all pass. We should hold off on merging until the changes are reflected in successive runs.

**Comment by mpalmi** [comment]
This is working on [enterprise](https://github.com/hashicorp/vault-enterprise/actions/runs/5514590794/jobs/10054064661?pr=4309). I'm going to keep running this until we get some balancing, but once this is merged, cache eviction shouldn't be as much of an issue.

File: .github/workflows/test-go.yml
Comment (miagilepner) [inline .github/workflows/test-go.yml:105]: what's the purpose of the cache key `read-go-test-reports-${{ github.run_number }}`? Is the intention that it won't match anything?

Comment (mpalmi) [reply .github/workflows/test-go.yml:105]: I believe this is intended to pick up the last cached read (if it exists). Otherwise, it will fall back to the `test-results` cache populated at the end of a successful run using `restore-keys`: https://github.com/hashicorp/vault/pull/21745/files#diff-6aac4b916cfd203239c824402c61b8d7257d7c849d4730b81041f05ac4569515R287

I pulled it from the example here: https://github.com/dnephin/infra/blob/main/.github/workflows/ci-core.yaml#L22

Comment (miagilepner) [reply .github/workflows/test-go.yml:105]: Ah ok. I don't think this key ever gets written to, since nothing else in the workflow has the key, and the [github.run_number](https://docs.github.com/en/actions/learn-github-actions/contexts) will be unique between runs. I guess since we have the restore-keys, it doesn't matter?

Comment (mpalmi) [reply .github/workflows/test-go.yml:105]: I think it gets populated every time we restore (at the beginning of the current run) from `go-test-...`. I'm not sure if you can restore from cache with only `restore-keys`. I need to think about this a bit more. I'll get back to you -- sorry for the uninformed speculation.

Comment (miagilepner) [reply .github/workflows/test-go.yml:105]: No worries, I also don't understand how it works 😅

Comment (mpalmi) [reply .github/workflows/test-go.yml:105]: Okay, I've done a bit of investigation and this is my understanding:

You cannot do a restore from cache without a `key` specified, so we need a key.

We don't want to put dirty state in the cache, since we don't have good logs until the end of the test run.

We restore the cache to a temporary new cache with the `read-` prefix, which as you pointed out is unique for each test run.

At the end, we pull the data down and store it to the persistent cache, which should be matched using `restore-keys` above.

We _may_ want to delete the read cache at the end using the cache API, or we can just leave read caches in and let the eviction take care of it.

File: .github/workflows/test-go.yml
Comment (miagilepner) [inline .github/workflows/test-go.yml:284]: this should use a commit sha, rather than v3

File: .github/workflows/test-go.yml
Comment (miagilepner) [inline .github/workflows/test-go.yml:289]: same here

File: .github/workflows/test-go.yml
Comment (miagilepner) [reply .github/workflows/test-go.yml:105]: might be worth adding a comment explaining why we use that key

File: .github/workflows/test-go.yml
Comment (mpalmi) [inline .github/workflows/test-go.yml:293]: Need 3.1.2, my bad.

**Review by miagilepner — APPROVED** [review]
LGTM! This is awesome!!!

File: .github/workflows/test-go.yml
Comment (mpalmi) [inline .github/workflows/test-go.yml:129]: We should probably just do this in its own step and store the results for re-use.

**Pull request merged**
